### PR TITLE
apps/shell, telnetd: dynamically allocate shell command buffer

### PR DESCRIPTION
**Pull request by sunghan-chang**
Currently TASH uses global array variable to save the command.
It sets the size from the config to get the maximum number of command.
But it's not good due to belows.
1. User should consider the maximum number of shell commands.
  When user enables more application or utils, then user should check it.
  It's much uncomfortable.
2. When user does not use it fully, it wastes memory.
3. Usually global variables are located in internal RAM which has fastest performance.
  But shell command itself does not require performance.
  And this takes memory even TASH is exited.

This commit changes that TASH dynamically allocates memory for the commands instead of
using global array variable.

### CODE REVIEW

File: apps/shell/tash_command.c [hunk 1]
```diff
@@ -638,18 +641,24 @@ bool tash_do_autocomplete(char *cmd, int *pos, bool double_tab)
...
+	if (tash_cmds_info.count == 0) {
+		new_cmd_buff = (struct tash_cmd_s *)malloc(sizeof(struct tash_cmd_s));
+	} else {
+		new_cmd_buff = (struct tash_cmd_s *)realloc(tash_cmds_info.cmd, sizeof(struct tash_cmd_s) * (tash_cmds_info.count + 1));
+	}
+
+	if (new_cmd_buff) {
+		tash_cmds_info.cmd = new_cmd_buff;
+	} else {
+		return -1; /* Memory Allocation Fail for new command */
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c:659]: How about printing the log we cannot install the cmd like before?

Comment (sunghan-chang) [reply apps/shell/tash_command.c:660]: Done

File: apps/shell/tash_command.c [hunk 2]
```diff
@@ -638,18 +641,25 @@ bool tash_do_autocomplete(char *cmd, int *pos, bool double_tab)
...
-		printf("Couldn't install cmd: (%s), Refer CONFIG_TASH_MAX_COMMANDS\n", str);
-		return -1;				/* MAX cmd count reached */
-	}
+	struct tash_cmd_s *new_cmd_buff;
 
 	/* Lock mutex */
 	pthread_mutex_lock(&tash_cmds_info.tmutex);
 
+	/* allocate the cmd buffer */
+	if (tash_cmds_info.count == 0) {
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c:650]: If command is already registered, we don't need to allocate the new.
So I think we'd better check the existence(line 664 to 669) before allocation or free the memory if already registered.

File: apps/shell/tash_command.c [hunk 3]
```diff
@@ -638,18 +641,25 @@ bool tash_do_autocomplete(char *cmd, int *pos, bool double_tab)
...
+		new_cmd_buff = (struct tash_cmd_s *)malloc(sizeof(struct tash_cmd_s));
+	} else {
+		new_cmd_buff = (struct tash_cmd_s *)realloc(tash_cmds_info.cmd, sizeof(struct tash_cmd_s) * (tash_cmds_info.count + 1));
+	}
+
+	if (new_cmd_buff) {
+		tash_cmds_info.cmd = new_cmd_buff;
+	} else {
+		printf("TASH: memory allocation fail to register (%s) command\n", str);
+		return -1; /* Memory Allocation Fail for new command */
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c:660]: In this case,  we should unlock the mutex(`tash_cmds_info.tmutex`).

File: apps/shell/tash_command.c [hunk 4]
```diff
@@ -638,24 +641,34 @@ bool tash_do_autocomplete(char *cmd, int *pos, bool double_tab)
...
 	pthread_mutex_lock(&tash_cmds_info.tmutex);
 
-	/* CHeck if cmd is already installed */
-	for (cmd_idx = 0; cmd_idx < tash_cmds_info.count; cmd_idx++) {
-		if (strncmp(str, tash_cmds_info.cmd[cmd_idx].str, TASH_CMD_MAXSTRLENGTH - 1) == 0) {
-			pthread_mutex_unlock(&tash_cmds_info.tmutex);
-			return -2;			/* CMD already installed */
+	/* allocate memory for new cmd */
+	if (tash_cmds_info.count == 0) {
+		new_cmd_buff = (struct tash_cmd_s *)malloc(sizeof(struct tash_cmd_s));
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c:651]: Please include `stdlib.h` for `malloc` and `realloc`.

**Review by jeongchanKim — APPROVED** [review]
LGTM

**Review by Taejun-Kwon — APPROVED** [review]

**Pull request merged**